lib/subroutine/fields.rb

```diff
@@ -0,0 +1,147 @@
+# frozen_string_literal: true
+
+require "active_support/concern"
+require "active_support/core_ext/object/duplicable"
+require "active_support/core_ext/hash/indifferent_access"
+require "active_support/core_ext/object/deep_dup"
+
+require "subroutine/type_caster"
+require "subroutine/association"
+
+module Subroutine
+  module Fields
+
+    extend ActiveSupport::Concern
+
+    included do
+      class_attribute :_fields
+      self._fields = {}
+      attr_reader :original_params
+      attr_reader :params, :defaults
+    end
+
+    class_methods do
+      # fields can be provided in the following way:
+      # field :field1, :field2
+      # field :field3, :field4, default: 'my default'
+      def field(*fields)
+        options = fields.extract_options!
+
+        fields.each do |f|
+          _field(f, options)
+        end
+      end
+      alias_method :fields, :field
+
+      def inputs_from(*things)
+        options = things.extract_options!
+        excepts = options.key?(:except) ? Array(options.delete(:except)) : nil
+        onlys = options.key?(:only) ? Array(options.delete(:only)) : nil
+
+        things.each do |thing|
+          thing._fields.each_pair do |field_name, opts|
+            next if excepts&.include?(field_name)
+            next if onlys && !onlys.include?(field_name)
+
+            if opts[:association]
+              include ::Subroutine::Association unless included_modules.include?(::Subroutine::Association)
+              association(field_name, opts)
+            else
+              field(field_name, opts)
+            end
+          end
+        end
+      end
+      alias_method :fields_from, :inputs_from
+
+      def respond_to_missing?(method_name, *args, &block)
+        ::Subroutine::TypeCaster.casters.key?(method_name.to_sym) || super
+      end
+
+      def method_missing(method_name, *args, &block)
```

Comment:
This means we can register in initializers and get the name of the caster readily available within an app. Example: a phone_number caster can be registered and we can use `phone_number` in ops as long as the initializer runs before the op is added.